### ISSUE TITLE
Make cfg immutable

### ISSUE THREAD

**Issue by erfannoury** [yassersouri/fandak#2]
What I have seen in a lot of projects that are based on config files is that the config file (and subsequently the config object in the project) becomes a form of global variable storage which is a code smell (I don't know the name of the anti-pattern). It's usually because it enable the user to have a mutable global state that can be modified anywhere in the code and thus makes following the logical progression of the program execution hard to follow (config can change anywhere in the code). Do you think it is possible to make config immutable in the project?

**Comment by yassersouri** [comment]
Thanks for the issue. This is true in my opinion.

YACS, provides a functionality to do this is a way. It is the [`freeze` function](https://github.com/rbgirshick/yacs/blob/aabe5bd54d193996ceedcff650cff10c38b1f90b/yacs/config.py#L248).

Since I want to update the config file in the main.py file usually, because I want to be able to change config files with command line arguments, I don't want to `freeze` the config file in the `get_default_config` function.

So I don't know what is the best thing to do. Should I do it [here](https://github.com/yassersouri/fandak/blob/05d3862330c6638f975de8542db15985526de076/examples/classification_mnist/main.py#L14)? What do you think is the best way?

**Comment by erfannoury** [comment]
Yes, I think that's a good place to freeze the config object. Basically the config object gets finalized after reading the config file and updating it based on input arguments. But this is just an opinion based on a short look that I had at the code. You have used the code extensively by now and know better.

**Comment by yassersouri** [comment]
Thank you now I freeze the config inside the `update_config` function.
